### PR TITLE
Eliminate ECMA_TRY_CATCH macros part I.

### PR DESCRIPTION
**Pull request by rerobika**
JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

**Comment by rerobika** [comment]
Note: the patch reduces the binary size by 592B on ARM without performance loss.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 1]
```diff
@@ -245,40 +246,40 @@ ecma_builtin_array_prototype_object_concat (const ecma_value_t args[], /**< argu
...
+    JERRY_ASSERT (ecma_is_value_empty (concat_value));
   }
-  else
+
+  ecma_value_t set_length_value = ecma_builtin_array_prototype_helper_set_length (new_array_p,
+                                                                                  ((ecma_number_t) new_length));
+  if (ECMA_IS_VALUE_ERROR (set_length_value))
   {
-    ecma_free_value (new_array);
+    return set_length_value;
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Freeing the `new_array_p` in case of error is not required?

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 2]
```diff
@@ -492,40 +492,37 @@ ecma_builtin_array_prototype_object_push (const ecma_value_t *argument_list_p, /
...
   {
-    /* 5.a */
-    ecma_value_t e_value = argument_list_p[index];
-
     /* 5.b */
     ecma_string_t *n_str_p = ecma_new_ecma_string_from_number (n);
 
-    ECMA_TRY_CATCH (put_value, ecma_op_object_put (obj_p, n_str_p, e_value, true), ret_value);
-    ECMA_FINALIZE (put_value);
+    ecma_value_t put_value = ecma_op_object_put (obj_p, n_str_p, argument_list_p[index], true);
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Can't we use `ecma_op_object_put_by_index` here?

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c (hunk 1 above)
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Thanks! Good catch.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c (hunk 2 above)
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: `ecma_op_object_put_by_index` is designed to handle only uint32_t indices, but at this point `ecma_number_t n = (ecma_number_t) length;` can overflow in the UINT32_MAX limit.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c (hunk 1 above)
Comment (galpeter) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Is this also means that there is no test for this? :)

Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Indeed we do not have test for it. The reason is that to trigger this branch we need to achieve that the newly constructed array's length should be greater than UINT32_MAX. We can create this kind of test but the routine iterates through the whole property chain to find the current indexed property which takes serious amount of time. So IMO we can assume that this deref is necessary but I'd not enclose a test case for this corner case.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 3]
```diff
@@ -642,59 +677,66 @@ ecma_builtin_array_prototype_object_shift (ecma_object_t *obj_p, /**< array obje
...
+    }
+
+    /* 7.b */
+    uint32_t to = k - 1;
 
     if (ecma_is_value_found (curr_value))
     {
       /* 7.d.i, 7.d.ii */
-      ECMA_TRY_CATCH (put_value, ecma_op_object_put (obj_p, to_str_p, curr_value, true), ret_value);
+      ecma_value_t put_value = ecma_op_object_put_by_index (obj_p, to, curr_value, true);
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: The "put" and "delete" parts are really similar, would it be possible to refactor it like this?
```c
ecma_value_t new_value;

if (ecma_value_is_found (curr_value))
{
    new_value = ecma_op_object_put_by_index (obj_p, to, curr_value, true);
    ecma_free_value (curr_value);
}
else
{
    new_value = ecma_op_object_delete_by_index (obj_p, to, true);
}

if (ECMA_IS_VALUE_ERROR (new_value))
{
    ecma_free_value (first_value);
    return new_value;
}
```

I'm okay having it in a followup PR.

Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Good idea, let me do it in a follow up patch.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 4]
```diff
@@ -1912,86 +1948,78 @@ ecma_builtin_array_reduce_from (ecma_value_t callbackfn, /**< routine's 1st argu
...
+    while (!k_present && index < len)
     {
       /* 8.b.i */
-      ecma_string_t *index_str_p = ecma_new_ecma_string_from_uint32 (start_from_left ? index
-                                                                                     : last_index - index);
       k_present = true;
 
       /* 8.b.ii-iii */
-      ECMA_TRY_CATCH (current_value, ecma_op_object_find (obj_p, index_str_p), ret_value);
+      ecma_value_t current_value = ecma_op_object_find_by_index (obj_p, start_from_left ? index : last_index - index);
```
[the remark this answers is not in the document]
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Good catch!

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 5]
```diff
@@ -1319,64 +1375,68 @@ ecma_builtin_array_prototype_object_unshift (const ecma_value_t args[], /**< arg
...
     ecma_string_t *to_str_p = ecma_new_ecma_string_from_number (new_idx);
 
-    /* 6.c */
-    ECMA_TRY_CATCH (get_value, ecma_op_object_find (obj_p, from_str_p), ret_value);
-
     if (ecma_is_value_found (get_value))
     {
       /* 6.d.i, 6.d.ii */
-      ECMA_TRY_CATCH (put_value, ecma_op_object_put (obj_p, to_str_p, get_value, true), ret_value);
+      ecma_value_t put_value = ecma_op_object_put (obj_p, to_str_p, get_value, true);
```
[the remark this answers is not in the document]
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: Indeed. I've applied your suggestion.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c [hunk 6]
```diff
@@ -1319,64 +1375,68 @@ ecma_builtin_array_prototype_object_unshift (const ecma_value_t args[], /**< arg
...
+        return put_value;
+      }
     }
     else
     {
       /* 6.e.i */
-      ECMA_TRY_CATCH (del_value, ecma_op_object_delete (obj_p, to_str_p, true), ret_value);
-      ECMA_FINALIZE (del_value);
-    }
+      ecma_value_t del_value = ecma_op_object_delete (obj_p, to_str_p, true);
```
[the remark this answers is not in the document]
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: ditto.

File: jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c (hunk 3 above)
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-array-prototype.c]: @galpeter I've already updated the PR with this idea.
